pystiche_papers/sanakoyeu_et_al_2018/_modules.py

```diff
@@ -184,3 +191,78 @@ def residual_block(channels: int) -> ResidualBlock:
     )
 
     return ResidualBlock(residual)
+
+
+def transformer_encoder(in_channels: int = 3,) -> SequentialEncoder:
+    r"""Encoder part of the :class:`Transformer` from :cite:`SKL+2018`.
+
+    Args:
+        in_channels: Number of channels in the input. Defaults to ``3``.
+
+    """
+    modules = (
+        nn.ReflectionPad2d(15),
+        ConvBlock(in_channels=in_channels, out_channels=32, kernel_size=3, stride=1),
+        ConvBlock(in_channels=32, out_channels=32, kernel_size=3, stride=2),
+        ConvBlock(in_channels=32, out_channels=64, kernel_size=3, stride=2),
+        ConvBlock(in_channels=64, out_channels=128, kernel_size=3, stride=2),
+        ConvBlock(in_channels=128, out_channels=256, kernel_size=3, stride=2),
+    )
+    return SequentialEncoder(modules)
+
+
+def transformer_decoder(
+    out_channels: int = 3, num_res_block: int = 9,
+) -> pystiche.SequentialModule:
+
+    modules: List[nn.Module] = []
+    for i in range(num_res_block):
+        modules.append(residual_block(256))
+
+    modules.append(UpsampleConvBlock(in_channels=256, out_channels=256, kernel_size=3))
+    modules.append(UpsampleConvBlock(in_channels=256, out_channels=128, kernel_size=3))
+    modules.append(UpsampleConvBlock(in_channels=128, out_channels=64, kernel_size=3))
+    modules.append(UpsampleConvBlock(in_channels=64, out_channels=32, kernel_size=3))
+    modules.append(nn.ReflectionPad2d(3))
+    modules.append(
+        conv(
+            in_channels=32,
+            out_channels=out_channels,
+            kernel_size=7,
+            stride=1,
+            padding="valid",
+        )
+    )
+    return pystiche.SequentialModule(*modules)
+
+
+class DecoderSigmoidOutput(nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return torch.sigmoid(input) * 2 - 1
+
+
+class Decoder(nn.Module):
+    r"""Decoder part of the :class:`Transformer` from :cite:`SKL+2018`."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.decoder = transformer_decoder()
+        self.output_module = DecoderSigmoidOutput()
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return cast(torch.Tensor, self.output_module(self.decoder(input)))
```

Comment:
Are you sure about this? `DecoderSigmoidOutput` maps the values to the interval `(-1.0, 1.0)`. Is there some other kind of post-processing?

Reply:
Yes, there is still a pre- and postprocessing of the images  to the interval `(-1.0, 1.0)`.

Reply:
In that case, please implement this in `_utils.py`

Reply:
I will do this in a following PR.

Reply:
Please do that here, since it is directly related.